### PR TITLE
build: Upgrade stale packages (fastapi, starlette, tensorflow, torch)

### PR DESCRIPTION
**Pull request by sakoush**
This is a change to upgrade deps of mlserver that was triggered by initially trying to update fastapi to a recent version. Due to some deps clashed we had to:
- remove py3.8 support: a few packages already require py3.9 as minimum version (e.g. [numpy](https://endoflife.date/numpy#python-support)  and [tensorflow](https://www.tensorflow.org/install/pip#software_requirements)). py3.8 EoL is [October 2024](https://devguide.python.org/versions/#supported-versions).
- update packages to more recent versions e.g. numpy `1.26.4`, tensorflow `2.14.1`  and torch `2.2.1`.
- update mypy to `1.8.0` to fix issues from starlette `0.36.3` upgrade
- fix tests from fastapi upgrade to `1.109.2` 

For a full list of changes check `pyproject.toml` and `poetry.lock`

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -55,14 +55,12 @@ uvicorn = "*"
 starlette-exporter = "*"
 py-grpc-prometheus = "*"
 aiokafka = "*"
-## Tritonclient has recently started requiring cuda-python, which doesn't work
-## in MacOS.
-## Adding a ceiling until this issue gets resolved:
-##    https://github.com/triton-inference-server/server/issues/6246
-tritonclient = {version = ">=2.24,<2.43", extras = ["http"]}
+tritonclient = "*"
```
[the remark this answers is not in the document]
Comment (sakoush) [reply pyproject.toml:58]: I feel this is just a stale comment that was not reflected in subsequent changes. yes adding a min version seems sensible.

File: pyproject.toml [hunk 2]
```diff
@@ -106,7 +104,7 @@ filelock = "^3.13.1"
 
 [tool.poetry.group.docker.dependencies]
 protobuf = "3.20.3"
-tensorflow = "~2.12.0"
+tensorflow = "<2.15"
```
[the remark this answers is not in the document]
Comment (sakoush) [reply pyproject.toml:107]: this version should be in sync with the version used in alibi and it is not yet migrated to 2.15+, might be just a version bump in alibi but I thought I will leave it for now as the ceiling defined there.

[the remark this answers is not in the document]
Comment (RafalSkolasinski) [reply pyproject.toml:108]: Thanks for clarifying! I have the feeling then it should only affect the alibi runtimes and not general compatibility of MLServer with tensorflow. 

If we put ceiling here explicitly if someone would want to build a custom runtime using tensorflow + mlserver (or extend mlserver-slim image) they would get artificial ceiling on the version of tensorflow used. 

In this sense, as long as mlserver itself is compatible with 2.x Tensorflow I would not put ceiling there. The effective lock file for alibi runtimes would resolve to version of tensorflow satisfying alibi requirements.

[the remark this answers is not in the document]
Comment (sakoush) [reply pyproject.toml:108]: I am not sure how the locks work in this case TBH. I think it is just a matter of being able to build the big mlserver image containing all runtimes. Maybe we can revisit how we can make it better in a follow up PR as there are note in the main `pyproject.toml` that assumes that the case (and for other deps from the different runtimes).

[the remark this answers is not in the document]
Comment (sakoush) [reply pyproject.toml:108]: specifically this section
```
[tool.poetry.group.all-runtimes-dev.dependencies]
## Dev dependencies from Alibi-Explain
requests-mock = "~1.11.0"
types-requests = "~2.28.11.5"

## Dev dependencies from Alibi-Detect
tensorflow = "<2.15"

## Dev dependencies from MLflow
torch = "2.2.1"
pytorch-lightning = "2.2.0.post0"
torchmetrics = "1.3.1"
torchvision = "0.17.1"
mlflow = "2.10.2"

## Dev dependencies from HuggingFace
transformers = ">=4.30,<5.0"
```

[the remark this answers is not in the document]
Comment (RafalSkolasinski) [reply pyproject.toml:108]: sure

**Comment by sakoush** [comment]
skipping flaky tests for now

**Pull request merged**